### ISSUE TITLE
app.py

### ISSUE THREAD

**Issue by AjibadeIB** [AjibadeIB/Resume#2]
import gradio as gr
import pandas as pd
import numpy as np
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

# Load data
fdf = pd.read_csv("movies_sample.csv")  # Ensure this file is uploaded in your Space

# Preprocess fields
fdf['release_date'] = pd.to_datetime(fdf['release_date'], errors='coerce')
fdf['release_year'] = fdf['release_date'].dt.year.fillna(0).astype(int)
fdf['poster_path'] = fdf['poster_path'].fillna('')
BASE_POSTER_URL = "https://image.tmdb.org/t/p/w185"

# Combine text fields
text_features = ['title','overview','genres','keywords','tagline','production_companies']
fdf['text'] = fdf[text_features].fillna('').agg(' '.join, axis=1)

# Load and configure model
model = SentenceTransformer('all-MiniLM-L6-v2')
model.max_seq_length = 256

# Batch encode texts
print("Encoding all movie texts in batches...")
batch_size = 512
texts = fdf['text'].tolist()
embeddings = []

for i in range(0, len(texts), batch_size):
    batch = texts[i:i+batch_size]
    emb = model.encode(batch, convert_to_numpy=True, normalize_embeddings=True)
    embeddings.append(emb)

embeddings = np.vstack(embeddings)

# --- Recommend function ---
def recommend(title_query, selected_title, genre_filter, year_filter):
    df_filtered = fdf.copy()

    if genre_filter:
        df_filtered = df_filtered[df_filtered['genres'].str.contains(genre_filter, case=False, na=False)]

    if year_filter:
        try:
            year = int(year_filter)
            df_filtered = df_filtered[df_filtered['release_year'] == year]
        except:
            return "Invalid year", [], [], None

    if selected_title:
        movie_index = df_filtered[df_filtered['title'] == selected_title].index[0]
    else:
        matches = df_filtered[df_filtered['title'].str.lower().str.contains(title_query.strip().lower(), na=False)]
        if matches.empty:
            return "❌ No movie found", [], [], None
        if len(matches) == 1:
            movie_index = matches.index[0]
        else:
            return gr.update(choices=matches['title'].tolist(), label="Multiple matches found. Pick one."), [], [], None

    # Similarity
    query_vector = embeddings[movie_index].reshape(1, -1)
    sim_scores = cosine_similarity(query_vector, embeddings).flatten()

    # Filtered indices
    filtered_indices = df_filtered.index.to_numpy()
    filtered_scores = sim_scores[filtered_indices]
    top_k = filtered_scores.argsort()[::-1]

    # Exclude itself
    top_k = [i for i in top_k if filtered_scores[i] < 0.999][:5]
    rec_indices = filtered_indices[top_k]

    rec_titles = fdf.loc[rec_indices, 'title'].tolist()
    rec_posters = [BASE_POSTER_URL + p if p else "" for p in fdf.loc[rec_indices, 'poster_path']]
    rec_df = fdf.loc[rec_indices, ['title', 'release_year', 'genres']]
    csv_path = "/tmp/recommendations.csv"
    rec_df.to_csv(csv_path, index=False)

    return f"Top 5 movies similar to **{fdf.loc[movie_index, 'title']}**:", rec_titles, rec_posters, csv_path

# --- Get matches for dropdown ---
def get_matches(query):
    matches = fdf[fdf['title'].str.lower().str.contains(query.strip().lower(), na=False)]
    return matches['title'].tolist() if not matches.empty else []

# --- Gradio UI ---
with gr.Blocks() as demo:
    gr.Markdown("# 🎬 Movie Recommender (BERT + Posters + Export)")

    with gr.Row():
        title_input = gr.Textbox(label="Enter Movie Title")
        find_btn = gr.Button("Find Matches")

    matched_dropdown = gr.Dropdown(choices=[], label="Select a Match (if multiple)", interactive=True)
    genre_input = gr.Textbox(label="Filter by Genre (optional)")
    year_input = gr.Textbox(label="Filter by Year (optional)")

    recommend_btn = gr.Button("Get Recommendations")

    output_text = gr.Markdown()
    output_list = gr.Dataframe(headers=["Recommended Movies"])
    poster_gallery = gr.Gallery(label="Poster Previews").style(grid=[5], height="auto")
    export_btn = gr.File(label="Download CSV")

    find_btn.click(fn=get_matches, inputs=title_input, outputs=matched_dropdown)

    recommend_btn.click(fn=recommend,
                        inputs=[title_input, matched_dropdown, genre_input, year_input],
                        outputs=[output_text, output_list, poster_gallery, export_btn])

demo. Launch()
